### ISSUE TITLE
compact missing from Int.NumberFormat.formatToParts 

### ISSUE THREAD

**Issue by mjpieters** [microsoft/TypeScript#42063]
**TypeScript Version:**  4.2.0-dev.20201221

**Search Terms:** `NumberFormat.formatToParts`, `compact`

**Expected behavior:**

`part: 'compact'` is a valid part type, as per [EMCA 402, section 12.1.7 (PartitionNotationSubPattern)](https://tc39.es/ecma402/#sec-partitionnotationsubpattern), step 4. - c. - iv. - 2.:

> iv. Else if _p_ is equal to **"compactSymbol"**, then
> 
>    1. *[...]*
>    2. Append a new Record { [[Type]]: **"compact"**, [[Value]]: _compactSymbol_ } as the last element of _result_.

Note that the [MDN documentation for `Intl.NumberFormat.prototype.formatToParts()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/formatToParts) doesn't mention this part type either; see mdn/content#516.

**Actual behavior:**

The [`Intl.NumberFormatPartTypes` type](https://github.com/microsoft/TypeScript/blob/6ea3885b8c6f9af64e536a707b31e2e15e3887d0/src/lib/esnext.intl.d.ts#L2) should include `"compact"`.

**Related Issues:**

* #42004 (*Intl.formatByParts() doesn't support the "unit" part type*, different part type, same area).

**Code**
```ts
const compactFormat = Intl.NumberFormat('en', {
  notation: 'compact',
  compactDisplay: 'short',
  style: 'decimal',
  maximumFractionDigits: 3,
})

// formatValue(42174812) => {formatted: '42.175', magnitude: 'M'}
function formatValue(value: number): {formatted: string, magnitude: string} {
  let magnitude = ''
  return {
    formatted: compactFormat
      .formatToParts(value)
      .reduce((formatted, part) => {
        switch (part.type) {
          case 'compact':
            magnitude = magnitude + part.value
            break
          default:
            formatted = formatted + part.value
        }
        return formatted
      }, ''),
    magnitude: magnitude
  }
}

console.log(formatValue(42174812))

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
const compactFormat = Intl.NumberFormat('en', {
    notation: 'compact',
    compactDisplay: 'short',
    style: 'decimal',
    maximumFractionDigits: 3,
});
// formatValue(42174812) => {formatted: '42.175', magnitude: 'M'}
function formatValue(value) {
    let magnitude = '';
    return {
        formatted: compactFormat
            .formatToParts(value)
            .reduce((formatted, part) => {
            switch (part.type) {
                case 'compact':
                    magnitude = magnitude + part.value;
                    break;
                default:
                    formatted = formatted + part.value;
            }
            return formatted;
        }, ''),
        magnitude: magnitude
    };
}
console.log(formatValue(42174812));

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "Latest",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?target=99&jsx=0#code/MYewdgzgLgBKC2AHAhsKAxEAnezYF4YBJMKAGwDoA5AV3gCMBTLTHPACgHJGxOAaGAG8AUDBhgQUPAEtwALhicEKNP1FwQSVFAAi0iIjLIAngs4QAFtihqx0Y2UZmAJo2DTcZWzFwAPD3ToWNqyYHoA5tJQEAoAzHzCAL4AlMLCAPTpMABm2LhQAGrIZDSM7AAsAEwAjADs5QAc1ZXJMPgAfEK5bFBQjM5mVRR1AKz8PsjhYFE0rmYAspyJwtk0YGihOXl4RSVlAG7FpQpgdExYyQqC3fl9AzDQWNJg4QK4UzNzD1BPL4lC6kcsHe0ygs0YbUUnHUWEYYKwYABYjENzwdwUym0rHy6mRMAoqKgABUQAAFZBYaLsQ57VJ4sQUWHOGjAMrsQl3AQoSmtDpI+l2ADuUWAFhg7G5UAoUGMiEYrREArxwGQEAhSk0KhsclxSrEIM+EMIBrBrhgAGoYJKKDTSrq9fRYcgANb2+mubLIGhkKA6vV4jn9SGB5wWq0UqW2xhusTLPWw+GIkNuxICTicZIJPEm8EKHOudTLZbCUCQECOChkEDhdnbQpHMpVOqNZrJVJAA)

**Comment by MartinJohns** [comment]
You are targetting ES3, but the 'compact' notation is an ES2020 feature.

**Comment by mjpieters** [comment]
> You are targetting ES3, but the 'compact' notation is an ES2020 feature.

I'm not. I don't know why the playground link, when opened, indeed shows 'ES3', but the config as included in the issue uses `"target": "Latest"` and locally I have `"target": "esnext"`.

I've also linked to the source code for `esnext.intl.d.ts` here, where you can see that the `NumberFormatPartTypes` type definition is clearly missing `"compact"` as an option.

**Comment by mjpieters** [comment]
I've found a few more missing types:

- `'unknown'`
- `'exponentSeparator'`
- `'exponentMinusSign'`
- `'exponentInteger'`

(also reported to MDN).

**Comment by andrewbranch** [comment]
Fixed by #42066
